web3/_utils/async_tools.py

```diff
@@ -0,0 +1,37 @@
+from functools import singledispatch, wraps
+import asyncio
+import inspect
+import types
+from typing import Any, Callable, Generator
+
+
+def _is_awaitable(co: Generator[Any, None, Any]) -> bool:
+    return inspect.isawaitable(co)
+
+
+@singledispatch
+def sync(co: Any):
+    raise TypeError('Called with unsupported argument: {}'.format(co))
+
+
+@sync.register(asyncio.Future)
+@sync.register(types.GeneratorType)
+def sync_co(co: Generator[Any, None, Any]) -> Any:
+    if not _is_awaitable(co):
+        raise TypeError('Called with unsupported argument: {}'.format(co))
+    return asyncio.get_event_loop().run_until_complete(co)
```

Comment:
It might be good to create a new event loop each time as I *think* that if the default event loop is already running this will fail (but I'm not sure about that).

Reply:
Sounds like a test we should run. Something like:

```py
async def test_async_calling_sync(event_loop, w3_sync):
  asyncio.set_event_loop(event_loop)  # I think this would set the test loop to the default event loop (which would catch the error piper mentioned), but it's worth dropping into pdb to be sure.
  actual_accounts = w3_sync.eth.accounts
  assert actual_accounts == expected_accounts
```

Reply:
It does fail.  sync() raises an error when Ive mistakenly had it called it in async tests.